project/src/main/java/com/google/sps/servlets/ContextDataServlet.java

```diff
@@ -0,0 +1,100 @@
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.common.collect.ImmutableList;
+
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.protobuf.ByteString;
+import java.util.HashMap;
+import java.nio.charset.StandardCharsets;
+
+import java.util.concurrent.*;
+
+/**
+ * Return the last 7 days of data to frontend for a given search query
+ */
+@WebServlet("/contextData")
+public class ContextDataServlet extends HttpServlet {
+  private static final Logger LOGGER = Logger.getLogger(ContextDataServlet.class.getName());
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Gson gson = new Gson();
+        InvestmentCalculator calc = new InvestmentCalculator();
+        String googleSearch = request.getParameter("search_term");
+
+        ImmutableList<Long> data = calc.getInvestmentDataIfExists(googleSearch);
+        if (data.isEmpty()) {
+            String date = calc.oneWeekBefore(calc.getLatestDate()).toString();
+
+            HashMap<String, String> arguments = new HashMap<>();
+            arguments.put("search", googleSearch);
+            arguments.put("date", date);
+
+            ByteString byteStr = ByteString.copyFrom(gson.toJson(arguments), StandardCharsets.UTF_8);
+            PubsubMessage pubsubApiMessage = PubsubMessage.newBuilder().setData(byteStr).build();
+            Publisher publisher = Publisher.newBuilder(
+                ProjectTopicName.of("google.com:sgonks-step272", "trendData")).build();
+            
+            try {
+                // Attempt to publish the message
+                publisher.publish(pubsubApiMessage).get();
+                // listen for data to be added to db
+                data = listenForDataOrTimeout(calc, data, googleSearch);
+            } catch (InterruptedException | ExecutionException e) {
+                LOGGER.log(Level.SEVERE, "Error publishing Pub/Sub message: " + e.getMessage(), e);
+            }
+        }
+
+        if (data.isEmpty()) {
+            // we still have no data - send timeout notice
+            LOGGER.log(Level.WARNING, "Timeout fetching investment data");
+            response.setContentType("application/html");
+            response.getWriter().println("<p>We were unable to fetch this data right now<p>");
+        } else {
+            // data retrieved from datastore, return data in json form
+            response.setContentType("application/json");
+            response.getWriter().println(gson.toJson(data));
+        }
+    }
+
+    /** 
+     * Check for data in datastore every 0.1 seconds until data exists or 15 seconds have elapsed
+     * Run this instead of waiting on cloud function return due to slight delay in datastore
+     */
```

Comment:
from our discussion is this your theory? 
... slight delay in datastore being written and becoming readable?

Reply:
I'm fairly certain that's the bug that I was getting before by waiting on a return.